### ISSUE TITLE
Alpaka function should accept non-const arguments by value or rvalue reference

### ISSUE THREAD

**Issue by fwyzard** [alpaka-group/alpaka#1923]
Various Alpaka functions that mutate their non-`const` argument take them by non-`const` reference:
  - `ALPAKA_FN_HOST static auto createTaskMemset(TView& view, std::uint8_t const& byte, TExtent const& extent)`
    takes `view` by non-`const` reference
  - `ALPAKA_FN_HOST auto enqueue(TQueue& queue, TTask&& task)`
    takes `queue` by non-`const` reference
  - _etc._

So far we have run into a couple of disadvantages:
  - it prevents the use of temporaries in the calls (because a non-`const` lvalue reference cannot bind to an rvalue):
    ```c++
    alpaka::memset(queue, alpaka::ViewPlainPtr<TDev, T, Dim1D, Idx>(data, device, Vec1D{extent}), 0x00);
    ```
    fails to compile because the `alpaka::ViewPlainPtr` rvalue cannot bind to a non-`const` lvalue

  - it requires our (`const`) objects to return a `Queue` by non-`const` reference:
    ```c++
    Queue& Holder::queue() const { return queue_; }
    ```
    to make it usable with `alpaka::enqueue(queue, ...)`:
    ```c++
    alpaka::memset(holder.queue(), buffer, 0x00);
    ```
    which in turn makes it possible to alter the content of the `queue_` member itself (_i.e._ make it point to a different queue altogether).

Unfortunately, it is not clear to us what would be the best strategy to address these problems.

On the one hand, the natural solution would seem to take the `queue`, `view`, _etc_. arguments by value: this is in line with the semantic of most Alpaka objects, and makes these calls work with values, rvalue references, `const` and non-`const` lvalue references.
The downside would be the creation of an extra temporary and the increment/decrement of the internal `shared_ptr` reference counter (unless the compiler is smart enough to optimise it all away).

On the other hand, it could be reasonable to take at least the `Queue` objects by `const` reference: functions that submit more work to a queue mutate its internal state, but not the state visible to the client code (the possible exception being the readiness of `Event`s, but those can change asynchronously from the client code anyway).

**Comment by fwyzard** [comment]
@makortel FYI

**Comment by psychocoderHPC** [comment]
Thanks for opening the issue. From the first read, I am not sure what will be the best solution to handle the issue.
We will discuss it in our developer meeting.

**Comment by fwyzard** [comment]
Unfortunately I cannot join today, let me know what you come up with.

**Comment by bernhardmgruber** [comment]
> So far we have run into a couple of disadvantages:
> 
> * it prevents the use of temporaries in the calls (because a non-`const` lvalue reference cannot bind to an rvalue):
>   ```c++
>   alpaka::memset(queue, alpaka::ViewPlainPtr<TDev, T, Dim1D, Idx>(data, device, Vec1D{extent}), 0x00);
>   ```
>   fails to compile because the `alpaka::ViewPlainPtr` rvalue cannot bind to a non-`const` lvalue

Right. And we had this problem in the past with `alpaka::memcpy`, where we changed the view argument to a forwarding reference. I think this is an API inconsistency and we should change it here as well.

```c++
void alpaka::memcpy(TQueue& queue, TViewDstFwd&& viewDst, TViewSrc const& viewSrc, TExtent const& extent)
```

> * it requires our (`const`) objects to return a `Queue` by non-`const` reference:
>   ```c++
>   Queue& Holder::queue() const { return queue_; }
>   ```
>   to make it usable with `alpaka::enqueue(queue, ...)`:
>   ```c++
>   alpaka::memset(holder.queue(), buffer, 0x00);
>   ```
>   which in turn makes it possible to alter the content of the `queue_` member itself (_i.e._ make it point to a different queue altogether).

I think this is a question of API design in alpaka. I am fine with enqueuing being a modifying operation on a queue. We discussed this case in the alpaka meeting today and the majority however seemed to view the queue object as a handle, without propagating constness. This is analoguous to the discussion on alpaka buffers in the past, where you pointed out that if an alpaka buffer is const, you could just take a copy and then mutate through the copy. With that philosophy, we could just as well make all queue paramters in alpaka APIs take const references.
